app/charges/controllers.py

```diff
@@ -269,6 +275,52 @@ def edit_charge(user, user_data):
             key == "status" or key == "paw_links" or key == "private" or key == "committee"):
             setattr(charge, key, user_data[key])
 
+        # Edit progress notes
+        if (key == "progress_notes"): 
+            current_notes = getattr(charge, "progress_notes") 
+
+            # Get the progress note info from user_data
+            note_id = int(user_data["progress_notes"]["id"])
+            note = str(user_data["progress_notes"]["note"])
+            date = str(user_data["progress_notes"]["date"])
+
+            # Edit the notes, and update the DB
+            current_notes.edit(note_id, [note, date, str(note_id)])
+            setattr(charge, "progress_notes", current_notes) 
+
+        #Delete progress_notes
+        if (key == "delete_note"): 
+            current_notes = getattr(charge, "progress_notes") 
+
+            note_id = int(user_data["delete_note"]["id"])
+
+            current_notes.pop(note_id)
+            setattr(charge, "progress_notes", current_notes)
+
+        # Create a new progess_note
+        if (key == 'add_note'):
+            # Get the current list of notes
+            current_notes = getattr(charge, "progress_notes")
+            if (current_notes == None):
+                current_notes = []
+
+            # Get the progress note info from user_data
+            note_id = str(user_data["add_note"]["id"])
+            note = str(user_data["add_note"]["note"])
+            date = str(user_data["add_note"]["date"])
+
+            # Create the new one from the user data
+            new_note = []
+            new_note.append(note)
+            new_note.append(date)
+            new_note.append(note_id)
```

Comment:
I feel like the client should not pass an id to create a note-- that should get handled by the backend.

Reply:
Good idea, that seems cleaner than what i was thinking, i forgot that the pop() method is using the index as a literal index. I think that if i change the pop() method to search for the index within the array, like a key, that should clear up your concerns. So the 'index' attribute will really just serve as a unique integer not an actual index.